control-plane/connect-inject/controllers/serviceaccount/serviceaccount_controller.go

```diff
@@ -0,0 +1,178 @@
+	"google.golang.org/protobuf/types/known/anypb"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/hashicorp/consul-k8s/control-plane/connect-inject/common"
+	"github.com/hashicorp/consul-k8s/control-plane/connect-inject/constants"
+	"github.com/hashicorp/consul-k8s/control-plane/consul"
+	"github.com/hashicorp/consul-k8s/control-plane/namespaces"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+	"github.com/hashicorp/go-multierror"
+)
+
+type Controller struct {
+	client.Client
+	// ConsulServerConnMgr is the watcher for the Consul server addresses used to create Consul API v2 clients.
+	ConsulServerConnMgr consul.ServerConnectionManager
+	// K8sNamespaceConfig manages allow/deny Kubernetes namespaces.
+	controllers.K8sNamespaceConfig
+	// ConsulNamespaceConfig manages settings related to Consul namespaces.
+	controllers.ConsulNamespaceConfig
+
+	Log logr.Logger
+
+	Scheme *runtime.Scheme
+	context.Context
+}
+
+func (r *Controller) Logger(name types.NamespacedName) logr.Logger {
+	return r.Log.WithValues("request", name)
+}
+
+func (r *Controller) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.ServiceAccount{}).
+		Complete(r)
+}
+
+// Reconcile reads the state of a ServiceAccount object for a Kubernetes namespace and reconciles the corresponding
+// Consul WorkloadIdentity.
+func (r *Controller) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var errs error
+	var serviceAccount corev1.ServiceAccount
+
+	// Ignore the request if the namespace of the service account is not allowed.
+	if common.ShouldIgnore(req.Namespace, r.DenyK8sNamespacesSet, r.AllowK8sNamespacesSet) {
+		return ctrl.Result{}, nil
+	}
+
+	// Create Consul resource service client for this reconcile.
+	resourceClient, err := consul.NewResourceServiceClient(r.ConsulServerConnMgr)
+	if err != nil {
+		r.Log.Error(err, "failed to create Consul resource client", "name", req.Name, "ns", req.Namespace)
+		return ctrl.Result{}, err
+	}
+
+	// If the ServiceAccount object has been deleted (and we get an IsNotFound error),
+	// we need to deregister that WorkloadIdentity from Consul.
+	err = r.Client.Get(ctx, req.NamespacedName, &serviceAccount)
+	if k8serrors.IsNotFound(err) {
+		err = r.deregisterWorkloadIdentity(ctx, resourceClient, req.Name, r.getConsulNamespace(req.Namespace), r.getConsulPartition())
+		return ctrl.Result{}, err
+	} else if err != nil {
+		r.Log.Error(err, "failed to get ServiceAccount", "name", req.Name, "ns", req.Namespace)
+		return ctrl.Result{}, err
+	}
+	r.Log.Info("retrieved ServiceAccount", "name", req.Name, "ns", req.Namespace)
+
+	// Ensure the WorkloadIdentity exists.
+	if err = r.registerWorkloadIdentity(ctx, resourceClient, &serviceAccount); err != nil {
```

Comment:
@DanStough I saw you mentioned adding some resource write backoffs in other controllers. Assuming the same approach should be used here, I can update once your work is merged or copy-able.

Reply:
Will either merge and push an update to that PR, or add to this PR, depending on which goes first.

Reply:
Added to follow-up ticket to ensure I don't forget